src/WebJobs.Script/Diagnostics/FunctionTraceWriterFactory.cs

```diff
@@ -1,33 +1,38 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the MIT License. See License.txt in the project root for license information.
 
+using System.Collections.Concurrent;
 using System.Diagnostics;
 using System.IO;
 using Microsoft.Azure.WebJobs.Host;
 
 namespace Microsoft.Azure.WebJobs.Script
 {
-    public sealed class FunctionTraceWriterFactory : ITraceWriterFactory
+    public sealed class FunctionTraceWriterFactory : IFunctionTraceWriterFactory
     {
-        private readonly string _functionName;
+        private readonly ConcurrentDictionary<string, TraceWriter> _writerCache = new ConcurrentDictionary<string, TraceWriter>();
         private readonly ScriptHostConfiguration _scriptHostConfig;
 
-        public FunctionTraceWriterFactory(string functionName, ScriptHostConfiguration scriptHostConfig)
+        public FunctionTraceWriterFactory(ScriptHostConfiguration scriptHostConfig)
         {
-            _functionName = functionName;
             _scriptHostConfig = scriptHostConfig;
         }
 
-        public TraceWriter Create()
+        public TraceWriter Create(string functionName)
         {
             if (_scriptHostConfig.FileLoggingMode != FileLoggingMode.Never)
             {
-                TraceLevel functionTraceLevel = _scriptHostConfig.HostConfig.Tracing.ConsoleLevel;
-                string logFilePath = Path.Combine(_scriptHostConfig.RootLogPath, "Function", _functionName);
-                return new FileTraceWriter(logFilePath, functionTraceLevel);
+                return _writerCache.GetOrAdd(functionName, f => CreateTraceWriter(_scriptHostConfig, f));
```

Comment:
K... so we do rely on the factory not outliving the writers with this implementation, right?